examples/conversation_sql/sql_utils.py

```diff
@@ -0,0 +1,308 @@
+          provide a brief explanation to the user explaining why their
+          question does not correspond to a SQL query within the
+          context of the schema. """
+
+    def format_target(self, example: dict) -> str:
+        """Format sql prompt"""
+        return self.format_question(example) + "\nSELECT "
+
+    def format_question(self, example: dict) -> str:
+        """Format question"""
+        sqls = get_sql_for_database(example["path_db"])
+        prompt_info = self.template_info.format("\n\n".join(sqls))
+        prompt_question = self.template_question.format(example["question"])
+        prompt_components = [prompt_info, prompt_question]
+        prompt = "\n\n".join(prompt_components)
+        return prompt
+
+    def describe_sql(self, example: dict) -> str:
+        """Describe SQL"""
+        sqls = get_sql_for_database(example["path_db"])
+        prompt_info = self.template_info.format("\n\n".join(sqls))
+        prompt_components = [
+            self.template_agent_prompt,
+            "DB schema info: ",
+            prompt_info,
+        ]
+        prompt = "\n\n".join(prompt_components)
+        return prompt
+
+    def is_sql_question(self, example: dict) -> str:
+        """whether the input is a sql question or not"""
+        sqls = get_sql_for_database(example["path_db"])
+        prompt_info = self.template_info.format("\n\n".join(sqls))
+        prompt_components = [
+            prompt_info,
+            self.is_sql_prompt,
+            example["question"],
+        ]
+        prompt = "\n\n".join(prompt_components)
+        return prompt
+
+
+class QuestionSqlExampleStyle:
+    """Provide QA pair as examples"""
+
+    def get_example_prefix(self) -> str:
+        """get example prefix"""
+        return (
+            "/* Some SQL examples are provided based on similar problems: */\n"
+        )
+
+    def format_example(self, example: dict) -> str:
+        """format example"""
+        template_qa = "/* Answer the following: {} */\n{}"
+        return template_qa.format(example["question"], example["query"])
+
+
+class EuclideanDistanceExampleSelector:
+    """Select similar sample question"""
+
+    def __init__(self) -> None:
+        self.train_json_path = "./database/train.json"
+        with open(self.train_json_path, "r", encoding="utf-8") as file:
+            self.train_json = json.load(file)
+        self.db_ids = [d["db_id"] for d in self.train_json]
+        self.train_questions = [_["question"] for _ in self.train_json]
+
+        self.SELECT_MODEL = "sentence-transformers/all-mpnet-base-v2"
+
+        from sentence_transformers import SentenceTransformer
+
+        self.bert_model = SentenceTransformer(self.SELECT_MODEL, device="cpu")
+        try:
+            self.train_embeddings = np.load("./.cache/train_embeddings.npy")
+        except FileNotFoundError:
+            self.train_embeddings = self.bert_model.encode(
```

Comment:
Maybe we can provide this embedding files directly, the same as the questions.

Reply:
The embedding file is 26mb and is a binary file format, I am not sure whether we should upload the file directly. Maybe it's better to provide a way to generate embeddings?